[13.0][FW] account_multicurrency_revaluation: forward port PR#149

JordiMForgeFlow opened the pull request:
Superseeds #183

alexey-pelykh committed:
[IMP] account_multicurrency_revaluation

hailangvn committed:
[IMP] account_multicurrency_revaluation: fix test, reverse moves

hailangvn committed:
[IMP] account_multicurrency_revaluation: black, isort, prettier

JordiMForgeFlow committed:
[IMP] account_multicurrency_revaluation: filter accounts by company

JordiMForgeFlow committed:
[IMP] account_multicurrency_revaluation: use active company instead of user company

JordiMForgeFlow committed:
[FIX] account_multicurrency_revaluation: fix report layout and field label

JordiMForgeFlow commented:
Hi @alexey-pelykh, @hailangvn and @bunhin

Any chance you can help me reviewing the changes? :)

hailangvn reviewed (approved):
LGTM. Thanks for the work.

MiquelRForgeFlow committed:
[FIX] account_multicurrency_revaluation: remove redundant default

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Start Revaluation Period

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Filter out non company accounts

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Start and End Date on report & Include Only Posted Entries

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Revaluation Date on Report

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Other improvements

alexey-pelykh reviewed (approved):
Code LGTM

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Readonly False for settings

GuillemCForgeFlow commented:
Hi @hailangvn , may you be able to review the changes I've added? :smile:

alexey-pelykh commented on account_multicurrency_revaluation/model/res_company.py:
```diff
@@ -77,4 +77,5 @@ class ResCompany(models.Model):
         string="Reversable Revaluations",
         help="Revaluations entries will be created " 'as "To Be Reversed".',
         default=True,
+        readonly=False,
```
why this is needed?

GuillemCForgeFlow replied:
Just to make sure that the field appears editable on the view. I could remove it and only add it to the field of the res_config file.

alexey-pelykh replied:
Yes, readonly=False should be done on related field, not here

GuillemCForgeFlow replied:
Great, I have applied the changes!

alexey-pelykh commented:
@GuillemCForgeFlow please rebase one more time to pickup new CI

adrienpeiffer commented:
/ocabot rebase

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Improve company and related fields

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Only include posted entries in the revaluation

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Replace flaged accounts for field

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: adapt tests

hailangvn commented on account_multicurrency_revaluation/model/account.py:
```diff
@@ -54,83 +78,109 @@ def _onchange_user_type_id(self):
 
-    def _revaluation_query(self, revaluation_date):
-
+    def _revaluation_query(self, start_date, revaluation_date):
```
IMHO we can enhance compatibility by using different method signature.
```suggestion
    def _revaluation_query(self, revaluation_date, start_date=None):
```

hailangvn commented on account_multicurrency_revaluation/model/account.py:
```diff
@@ -54,83 +78,109 @@ def _onchange_user_type_id(self):
-        # compute for each account the balance/debit/credit from the move lines
-        query, params = self._revaluation_query(revaluation_date)
+    def compute_revaluations(self, start_date, revaluation_date):
+        query, params = self._revaluation_query(start_date, revaluation_date)
```
Especially for public method like this.
```suggestion
    def compute_revaluations(self, revaluation_date, start_date=None):
        query, params = self._revaluation_query(revaluation_date, start_date)
```

hailangvn commented on account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py:
```diff
@@ -13,26 +15,61 @@ class UnrealizedCurrencyReportPrinter(models.TransientModel):
     )
+    start_date = fields.Date(
+        string="Start Date",
+        help="The report will print from this Date, all the revaluated entries"
```
I think space is needed at end of this line (or at beginning of next line).
```suggestion
        help="The report will print from this Date, all the revaluated entries "
```

hailangvn commented on account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py:
```diff
@@ -310,83 +277,102 @@ def revaluate_currency(self):
 
-        # Get balance sums
-        account_sums = account_ids.compute_revaluations(self.revaluation_date)
+        revaluations = account_ids.compute_revaluations(
+            self.start_date, self.revaluation_date
+        )
```
```suggestion
        revaluations = account_ids.compute_revaluations(
            self.revaluation_date, self.start_date
        )
```

hailangvn reviewed (commented):
Thanks for the work. It looks good except some minor points to help improving experience of users and developers. And as we are on this, please consider to update below help string and its translations.

* [account_multicurrency_revaluation/model/res_company.py lines 84-85](https://github.com/OCA/account-closing/pull/200/files#diff-2c16930dd0dbc7275aa3510f3e0dce93fcf838e476db03cfa7db71ac594bd5f8R84-R85) add a space at end/beginning of a line.

```
        help="If marked, the entries created in the process will be "
        "posted automatically.",
```

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation
Adapt helps to suggestions and adapt method declarartion to improve the experience of users/developers.

GuillemCForgeFlow commented:
Hi @hailangvn , you're right! I have added you're suggestions :+1:

hailangvn commented on account_multicurrency_revaluation/model/res_company.py:
```diff
@@ -76,13 +76,13 @@ class ResCompany(models.Model):
         string="Auto Post Created Entries",
         help="If marked, the entries created in the process will be"
-        "posted automatically.",
+        "posted automatically. ",
```
The space is needed to separate `be` and `posted` so below is the suggested change.
```suggestion
        " posted automatically.",
```

hailangvn commented on account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py:
```diff
@@ -21,12 +21,13 @@ class UnrealizedCurrencyReportPrinter(models.TransientModel):
         string="Start Date",
         help="The report will print from this Date, all the revaluated entries"
-        "created from this date. The default value will be the first day of the month.",
+        "created from this date. The default value will be the first day of the month."
+        " ",
```
The space is needed to separate `entries` with `created` so below is the suggested change.
```suggestion
        " created from this date. The default value will be the first day of the month."
```

hailangvn commented on account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py:
```diff
@@ -21,12 +21,13 @@ class UnrealizedCurrencyReportPrinter(models.TransientModel):
     end_date = fields.Date(
         string="End Date",
-        help="The report will print till this Date. The default value will be today.",
+        help="The report will print till this Date. The default value will be today. ",
```
This one only has one line so no need space at end.
```suggestion
        help="The report will print till this Date. The default value will be today.",
```

hailangvn commented on account_multicurrency_revaluation/model/res_company.py:
```diff
@@ -76,13 +76,13 @@ class ResCompany(models.Model):
         string="Flag Revaluation Entries",
         help="If marked, the entries created in the process will be flaged "
-        "to be able to filter them out later.",
+        "to be able to filter them out later. ",
```
The previous line has space at end already so no need to change this line.
```suggestion
        "to be able to filter them out later.",
```

hailangvn commented on account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py:
```diff
@@ -56,7 +56,7 @@ def _get_default_revaluation_account_ids(self):
         string="Journal",
         domain=[("type", "=", "general")],
-        help="You can set the default journal in company settings.",
+        help="You can set the default journal in company settings. ",
```
This one only has one line so no need space.
```suggestion
        help="You can set the default journal in company settings.",
```

hailangvn commented on account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py:
```diff
@@ -72,7 +72,7 @@ def _get_default_revaluation_account_ids(self):
         comodel_name="account.account",
         string="Revaluation Accounts",
-        help="Accounts that will be revaluated",
+        help="Accounts that will be revaluated. ",
```
This one only has one line so no need space.
```suggestion
        help="Accounts that will be revaluated.",
```

hailangvn reviewed (approved):
Appreciate your effort for this. Sorry that I was not specific enough in previous comments. So the minimum changes are the two spaces to separate words. Other changes are just nitpicking on the way.

GuillemCForgeFlow commented:
Hi @hailangvn , now I got you! I've applied the suggestions :+1:

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Remove currency in line
Removed as it may lead to problems just because the entry may have the currency from the company, and the lines having a different currency. Then we will not be able to post the entry.

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Reverse moves
Wizard to be able to create the reverse moves given an start date and end date. All revaluated entries during the period will be included and upon the validation, the reverse moves will be created. Also you can determine a posting date which will be the date assigned to all reverse moves created.

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Always flag revaluated entries
Remove the setting that allowed the user to possibly flag the revaluated entries created during the main process. Always flag them instead to be used later in the reverse wizard.

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Small improvements
Make both fields readonly as they will be only used in the logic and to filter, the user will not modify them. Also, limit entries appearing in reverse wizard.

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Line origin
Adds the origin line from where the revaluated line is created. Also, improves the query to get the open items.

GuillemCForgeFlow committed:
[FIX]account_multicurrency_revaluation: Use all open items
Before, the lines were overwriten, so if there were multiple lines for the same account, partner and currency, the last open item was the only one used. Now the amounts are acumulated into one line.

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation
Two-sided relation between lines created and original ones

hailangvn reviewed (approved):
LGTM. Thank you.

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Adjust values if partial reconciled items
If the origin line that will be revaluated has partially reconciled items, adjust the values to the real ones.

GuillemCForgeFlow committed:
[IMP]account_multicurrency_revaluation: Replace label for span

GuillemCForgeFlow committed:
[FIX]account_multicurrency_revaluation: Fix tests

MiquelRForgeFlow reviewed (approved):
I don't know if it's okey having that many commits instead of squashing... but LGTM

AaronHForgeFlow reviewed (approved):
LGTM

AaronHForgeFlow commented:
/ocabot merge minor
